content/courses/Ada_For_The_C_Embedded_Developer/chapters/06_Translation.rst

```diff
@@ -725,9 +725,467 @@ above could also be literally translated to:
+    {
+        int i, j;
+        char *c = (char *)bits;
+
+        assert(sizeof(char) == 1);
+
+        printf("Bits: ");
+        for (i = 0; i < len / (sizeof(char) * 8); i++)
+        {
+            for (j = 0; j < sizeof(char) * 8; j++)
+            {
+                printf("%d", c[i] >> j & 1);
+            }
+        }
+        printf("\n");
+    }
+
+    !bitfield_serialization.c
+    #include <stdio.h>
+
+    #include "my_recs.h"
+    #include "serializer.h"
+
+    int main(int argc, const char * argv[])
+    {
+        rec r = {5, "abc"};
+
+        transmit(&r, sizeof(r) * 8);
+    }
+
+Similarly, we can write a subprogram that converts a bit-field |mdash| which
+may have been received as a bitstream |mdash| to a specific type. We can add a
+:ada:`To_Rec` subprogram to the :ada:`My_Recs` package to convert a bit-field
+to the :ada:`Rec` type. This can be used to convert a bitstream that we
+received into the actual data type representation.
+
+As you know, we may write the :ada:`To_Rec` subprogram as a procedure or as a
+function. Since we need to use slightly different strategies for the
+implementation, the following example has both versions of :ada:`To_Rec`.
+
+This is the updated code for the :ada:`My_Recs` package and the :ada:`Main`
+procedure:
+
+[Ada]
+
+.. code:: ada run_button project=Courses.Ada_For_C_Embedded_Dev.Translation.Bitfield_Deserialization_Ada
+
+    package Serializer is
+
+       type Bit_Field is array (Natural range <>) of Boolean with Pack;
+
+       procedure Transmit (B : Bit_Field);
+
+    end Serializer;
+
+    with Ada.Text_IO; use Ada.Text_IO;
+
+    package body Serializer is
+
+       procedure Transmit (B : Bit_Field) is
+
+          procedure Show_Bit (V : Boolean) is
+          begin
+             case V is
+                when False => Put ("0");
+                when True  => Put ("1");
+             end case;
+          end Show_Bit;
+
+       begin
+          Put ("Bits: ");
+          for I in B'Range loop
+             Show_Bit (B (I));
+          end loop;
+          New_Line;
+       end Transmit;
+
+    end Serializer;
+
+    with Serializer;  use Serializer;
+
+    package My_Recs is
+
+       type Rec is record
+          V : Integer;
+          S : String (1 .. 3);
+       end record;
+
+       procedure To_Rec (B :     Bit_Field;
+                         R : out Rec);
+
+       function To_Rec (B : Bit_Field) return Rec;
+
+       procedure Display (R : Rec);
+
+    end My_Recs;
+
+    with Ada.Text_IO; use Ada.Text_IO;
+
+    package body My_Recs is
+
+       procedure To_Rec (B :     Bit_Field;
+                         R : out Rec) is
+          B_R : Bit_Field (0 .. R'Size - 1) with Address => R'Address;
+       begin
+          --  Assigning data from bit-field parameter B to bit-field
+          --  representation of output parameter.
+          B_R := B;
+       end To_Rec;
+
+       function To_Rec (B : Bit_Field) return Rec is
+          R   : Rec;
```

Comment:
Same as before, I would just give to R the address of B and return R directly. In this particular case, you're saving an extra copy :-) (which arguably would probably be optimized at high levels).